lib/internal/modules/cjs/loader.js

```diff
@@ -357,6 +359,126 @@ function findLongestRegisteredExtension(filename) {
   return '.js';
 }
 
+function resolveBasePath(basePath, exts, isMain, trailingSlash, request) {
+  let filename;
+
+  const rc = stat(basePath);
+  if (!trailingSlash) {
+    if (rc === 0) {  // File.
+      if (!isMain) {
+        if (preserveSymlinks) {
+          filename = path.resolve(basePath);
+        } else {
+          filename = toRealPath(basePath);
+        }
+      } else if (preserveSymlinksMain) {
+        // For the main module, we use the preserveSymlinksMain flag instead
+        // mainly for backward compatibility, as the preserveSymlinks flag
+        // historically has not applied to the main module.  Most likely this
+        // was intended to keep .bin/ binaries working, as following those
+        // symlinks is usually required for the imports in the corresponding
+        // files to resolve; that said, in some use cases following symlinks
+        // causes bigger problems which is why the preserveSymlinksMain option
+        // is needed.
+        filename = path.resolve(basePath);
+      } else {
+        filename = toRealPath(basePath);
+      }
+    }
+
+    if (!filename) {
+      // Try it with each of the extensions
+      if (exts === undefined)
+        exts = Object.keys(Module._extensions);
+      filename = tryExtensions(basePath, exts, isMain);
+    }
+  }
+
+  if (!filename && rc === 1) {  // Directory.
+    // try it with each of the extensions at "index"
+    if (exts === undefined)
+      exts = Object.keys(Module._extensions);
+    filename = tryPackage(basePath, exts, isMain, request);
+  }
+
+  return filename;
+}
+
+// This only applies to requests of a specific form:
+// 1. @
+// 2. @/.*
+const SELF_PATTERN = /^@(\/.*)?$/;
+function trySelf(paths, exts, isMain, trailingSlash, request) {
+  if (!experimentalExports) {
+    return false;
+  }
+
+  const [, expansion] =
+    StringPrototype.match(request, SELF_PATTERN) || [];
+
+  if (!expansion) {
+    return false;
+  }
+
+  for (var i = 0; i < paths.length; i++) {
+    // Each path is the node_modules sibling of a potential package.json dir
+    const basePath = path.resolve(paths[i], '..');
+
+    // Skip directories that don't have a package.json
+    const pkg = readPackageRaw(basePath);
+    if (!pkg) continue;
+
+    if (expansion) {
+      // Use exports
+      const fromExports = applyExports(basePath, expansion);
+      if (!fromExports) return false;
+      return resolveBasePath(fromExports, exts, isMain, trailingSlash, request);
+    } else {
+      // Use main field
+      return tryPackage(basePath, exts, isMain, request);
+    }
+  }
```

Comment:
Could we just do a direct package scope lookup here instead. Having a dependence on paths seems like it would also introduce deviation between the ESM and CJS implementations so I would prefer an approach that works equally for both.